### PR TITLE
Place picker refactor and tests.

### PR DESCRIPTION
**Pull request by kejmil01**

### CODE REVIEW

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/GpsEstablisher.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    public GpsEstablisher(
+            LocationManager locationManager,
+            GoogleLocationApiEstablisher googleLocationApiEstablisher,
+            Activity activity) {
+        this.googleLocationApiEstablisher = googleLocationApiEstablisher;
+        this.locationManager = locationManager;
+        this.weakActivityReference = new WeakReference<>(activity);
+    }
+
+    int getGpsRequestCode() {
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/GpsEstablisher.java]: Redundant

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -9,6 +9,7 @@
     <string name="error_message">There are some problems with editing this tag</string>
     <string name="enter_value_hint">Enter value</string>
     <string name="google_services_error">Google Play Services is not available.</string>
+    <string name="google_api_connection_error">Could not connect with google api.</string>
```
Comment (BartSoftware) [inline app/src/main/res/values/strings.xml]: Could not connect to google api.

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GoogleLocationApiEstablisherTest.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+        //When
+        boolean apiConnected = establisher.isApiConnected();
+
+        //Then
+        assertEquals(false, apiConnected);
+    }
+
+    @Test
+    public void whenGoogleApiBuildAndConnected_ApiIsConnected() {
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GoogleLocationApiEstablisherTest.java]: whenGoogleApiIsBuiltAndConnected_ApiIsConnected

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GoogleLocationApiEstablisherTest.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+        //When
+        boolean apiConnected = establisher.isApiConnected();
+
+        //Then
+        assertEquals(true, apiConnected);
+    }
+
+    @Test
+    public void whenConnectionRequested_ApiClientBuild() {
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GoogleLocationApiEstablisherTest.java]: whenConnectionIsRequested_ApiClientIsBuilt

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GoogleLocationApiEstablisherTest.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        //When
+        establisher.requestConnection();
+
+        //Then
+        assertNotNull(establisher.googleApiClient);
+    }
+
+    @Test
+    public void whenOnConnectedInvoked_ListenerOnSuccessCalled() {
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GoogleLocationApiEstablisherTest.java]: whenOnConnectedIsInvoked_ListenerOnSuccessIsCalled

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GpsEstablisherTest.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        callbackCaptor.getValue().onResult(result);
+
+        //Then
+        verify(statusListener).onGpsEstablished();
+    }
+
+    @Test
+    public void whenLocationStatusResolutionRequiredAndThereIsActivity_startResolutionForResultCalled()
+            throws IntentSender.SendIntentException {
+        //given
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GpsEstablisherTest.java]: //Given

**Review by BartSoftware — CHANGES_REQUESTED** [review]

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GoogleLocationApiEstablisherTest.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+        //When
+        connectionCallbacks.onConnected(null);
+
+        //Then
+        verify(listener).onSuccess();
+    }
+
+    @Test
+    public void whenOnConnectedIsInvokedAndThereIsNoListener_ErrorisNotThrown() {
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GoogleLocationApiEstablisherTest.java]: whenOnConnectedIsInvokedAndThereIsNoListener_ErrorIsNotThrown

File: app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GpsEstablisherTest.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        boolean gpsEstablished = gpsEstablisher.isGpsEstablished();
+
+        //Then
+        assertTrue(gpsEstablished);
+    }
+
+    @Test
+    public void whenLocationStatusIsSuccess_GpsEstablishedIsCalled() {
+        //Given
+        GpsEstablisher.StatusListener statusListener = mock(GpsEstablisher.StatusListener.class);
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/mvvm/edition/GpsEstablisherTest.java:55]: BoilerPlate code, please move it for some method

**Review by BartSoftware — CHANGES_REQUESTED** [review]

**Review by BartSoftware — APPROVED** [review]

**Review by michal1990 — APPROVED** [review]

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/EditActivity.java [hunk 9]
```diff
@@ -89,23 +90,30 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
         if (savedInstanceState != null) {
             tag = savedInstanceState.getParcelable(TAG_MODEL);
         }
+        editActivityHelper = new EditActivityHelper(this);
         prepareLocationHandling();
     }
 
+    @Override
+    protected void onDestroy() {
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/EditActivity.java]: onDestroy() is not a good idea, its not called every time, much better is onStop()

**Review by BartSoftware — CHANGES_REQUESTED** [review]

**Review by BartSoftware — APPROVED** [review]

**Pull request merged**